[FEATURE]: Project deployment

Maximus7474 opened the pull request:
### 📄 **Pull Request Description**
<!-- Provide a brief summary of your pull request. What does it achieve or modify? -->
This PR introduces a build script for the server backend to bundle all dependencies together into one file using esbuild.
Also reworked the release workflow to prepare the built files and create release (until a new strategy is thought of or is defined).

---

### 📝 **Major Changes**
<!-- List the key changes this PR introduces. For example: -->
- Reworked the release workflow to build the project and create a release with the built files.
- Altered the backend to only use environment variables for all configuration related information
  - DB connection credentials
  - JWT secret
  - etc...

---

### ✅ **Type of Modifications**
<!-- Delete the lines that don't apply -->
- 🍃 New Feature  
- ♻️ Refactor  
- 🎒 Other (please specify):  CI

---

### 🔗 **Related Issues**
<!-- Link to any related issues or pull requests, if applicable. Use: Fixes #123 or Closes #456 -->
- Addresses discussion in #23 

---

### 📊 **Testing Steps**
<!-- Outline how your changes can be tested. Include steps, environments, or edge cases considered. -->
1. run `npm run build` in `src/client/` and `src/server/`
2. create a .env file from the template in `src/build/server/`
3. execute `cd src/build/server/ node server.js` to launch project

---

### 🛑 **Checklist Before Merging**
- [x] Code follows the project's coding standards.
- [ ] Appropriate tests have been added or updated.
- [ ] Documentation updated where necessary.
- [x] PR title and description are clear and concise.
- [x] All checks pass (CI/CD, linting).

---

### 💬 **Additional Context**
<!-- Add any other context or information about the pull request here. -->

Maximus7474 committed:
refactor(server): set JWT Secret to be defined as environment variable

Maximus7474 committed:
feat(server/config): added validation of JWT Secret in .env

Maximus7474 committed:
rf(server/models): use environment variables and not config.json for DB auth

Maximus7474 committed:
refactor(server/server.js): don't verify for config.json & remove it from git ignore

Maximus7474 committed:
feat(server): esbuild script

Maximus7474 committed:
refactor(client/vite.config): defined custom outDir

Maximus7474 committed:
refactor(server/models): set dialectModule as "pg" & install as dependancy for build

Maximus7474 committed:
refactor(workflow/release): integrated build steps, update package versions and finalize release.zip

Maximus7474 committed:
rm: unused package.json in src/

Brightiya commented on src/server/.env.template:
in:( DB_USERNAME=postgres), the database username I used was 'bright' not postgres. 
Is this: ("dialect": "postgres"), not  needed anymore?

Maximus7474 replied:
The `.env.template` is a template file credentials for DB will be different from one user to the next.
On my end I'm using the default `postgre` account name, a different password and port for it.

Relying on a static json file to store these isn't the best solution, especially for a public repo and where many maintainers can come to work with. The JWT secret, DB credentials for production should only be known to the sys. admin and stored on the server hosting the project.

The dialect has been hardcoded in to avoid possible issues in case someone sets it to something else.

Brightiya replied:
Okay.

Maximus7474 committed:
Merge branch 'main' into production-deployment

Maximus7474 commented:
Will update setup documentation

Maximus7474 committed:
fix(server.js): serve client files if running in production
-> so that when connecting to the webserver the client downloads all relevant files

Maximus7474 committed:
feat(server/package): build script

Maximus7474 committed:
fix(server/config): not veryfing PORT from .env file

Maximus7474 committed:
fix(server/server.js): require hierachy for config variables

Maximus7474 commented:
Just thought of something when reviewing the code, I'll need your opinion @Brightiya .

Currently the DB models are dynamically imported into the Sequelize class object, this is good however it doesn't play well with `esbuild`, what it will do is ignore those files when building and not therefor they won't exist on the built project causing major issues.

I'm not sure what the best option is so I'll list them below:
- Remove the esbuild bundling operation and have the project run using node_modules/ for all it's dependancies
  - ✅ Pros: No issues with having to write code with bundling limitations in mind
  - ❌Cons: Larger deployment size, slower startup time, more complex deployment (if deploying to another plateform)
- Dynamically list files using `fs.readdirSync` and specify in the build script that it should not bundle the models but transfer them
  - ✅Pros: No need to add files manually to the sequelize initialisation
  - ❌Cons: possible issues implementing it, can take a bit of time to fine tune
- Hard code the imports, using `const Users = require("user.js");`  for each of the models
  - ✅Pros: Allow the use of esbuild without any hassle of issues
  - ❌Cons: Each model needs to be added manually to `server/models/index.js`

Brightiya commented:
@Maximus7474 since there is going to be issues using esbuild with Sequel, then let's go for hard coding the import for now coz we don't have much time again: Let's get the app running(at least the homepage and user dashboard) so that we can add  further features.

Maximus7474 commented:
> @Maximus7474 since there is going to be issues using esbuild with Sequel, then let's go for hard coding the import for now coz we don't have much time again: Let's get the app running(at least the homepage and user dashboard) so that we can add  further features. 

Sounds good, will integrate it tonight. I'll drop a test procédure so you can review and validate the PR

Maximus7474 committed:
refactor(server/models): hard coded model require paths to work with esbuild

Maximus7474 committed:
refactor(server): require path for sequelize/db module

Maximus7474 committed:
tweak(server/user.controller): destructure db import as in other controllers

Maximus7474 committed:
fix(server/server.js): missing `path` import

Maximus7474 committed:
chore(documentation/server): update set up & basic usage

Maximus7474 committed:
documentation: added build-and-deploy guide

Maximus7474 commented:
@Brightiya PR ready for review, you can check the deployment steps in [build-and-deploy](https://github.com/ManahilMuhammad/asp-team-20/blob/940d483771c4d3b54c89746f97d5f80f4cf18bec/documentation/setup_guides/build-and-deploy.md) guide

Project should:
- [ ] Build without issues
- [ ] Be usable the same as in development & production environment

Brightiya reviewed (approved):
PR will be merged and tested.
